File: sdks/go/cross_sdk_test.go
// Cross-SDK consistency tests — verify the Go SDK parses SSE fixtures
// identically to the Python and TypeScript SDKs. The exact same SSE byte
// strings are used in sdks/python/tests/test_cross_sdk.py and
// sdks/typescript/src/cross-sdk.test.ts.

package langdag

import (
	"io"
	"strings"
	"testing"
)

// Canonical SSE fixtures — byte-for-byte identical across all 3 SDKs.
const (
	// Normal happy path: start → delta → delta → done
	fixtureNormal = "event: start\ndata: {}\n\n" +
		"event: delta\ndata: {\"content\":\"Hello \"}\n\n" +
		"event: delta\ndata: {\"content\":\"world!\"}\n\n" +
		"event: done\ndata: {\"node_id\":\"test-node-1\"}\n\n"

	// Error mid-stream: start → delta → error
	fixtureErrorMidStream = "event: start\ndata: {}\n\n" +
		"event: delta\ndata: {\"content\":\"partial \"}\n\n" +
		"event: error\ndata: provider crashed\n\n"

	// Multi-line error: start → error with 3 data lines
	fixtureMultiLineError = "event: start\ndata: {}\n\n" +
		"event: error\ndata: line one\ndata: line two\ndata: line three\n\n"

	// Error-only: no start, no content
	fixtureErrorOnly = "event: error\ndata: unauthorized\n\n"
)

func streamFromFixture(fixture string) *Stream {
	return newStream(io.NopCloser(strings.NewReader(fixture)), nil)
}

func drainEvents(s *Stream) []SSEEvent {
	var events []SSEEvent
	for e := range s.Events() {
		events = append(events, e)
	}
	return events
}

func TestCrossSDK_NormalFlow(t *testing.T) {
	s := streamFromFixture(fixtureNormal)
	events := drainEvents(s)

	if len(events) != 4 {
		t.Fatalf("expected 4 events, got %d", len(events))
	}
	if events[0].Type != "start" {
		t.Errorf("event 0: expected start, got %s", events[0].Type)
	}
	if events[1].Type != "delta" || events[1].Content != "Hello " {
		t.Errorf("event 1: expected delta 'Hello ', got %s %q", events[1].Type, events[1].Content)
	}
	if events[2].Type != "delta" || events[2].Content != "world!" {
		t.Errorf("event 2: expected delta 'world!', got %s %q", events[2].Type, events[2].Content)
	}
	if events[3].Type != "done" || events[3].NodeID != "test-node-1" {
		t.Errorf("event 3: expected done/test-node-1, got %s/%s", events[3].Type, events[3].NodeID)
	}

	if s.Content() != "Hello world!" {
		t.Errorf("content: expected 'Hello world!', got %q", s.Content())
	}
	if s.Err() != nil {
		t.Errorf("expected no error, got %v", s.Err())
	}

	node, err := s.Node()
	if err != nil {
		t.Fatalf("Node() error: %v", err)
	}
	if node.ID != "test-node-1" {
		t.Errorf("expected node test-node-1, got %s", node.ID)
	}
}

func TestCrossSDK_ErrorMidStream(t *testing.T) {
	s := streamFromFixture(fixtureErrorMidStream)
	events := drainEvents(s)

	if len(events) != 3 {
		t.Fatalf("expected 3 events, got %d", len(events))
	}
	if events[0].Type != "start" {
		t.Errorf("event 0: expected start, got %s", events[0].Type)
	}
	if events[1].Type != "delta" || events[1].Content != "partial " {
		t.Errorf("event 1: expected delta 'partial ', got %s %q", events[1].Type, events[1].Content)
	}
	if events[2].Type != "error" || events[2].Error != "provider crashed" {
		t.Errorf("event 2: expected error 'provider crashed', got %s %q", events[2].Type, events[2].Error)
	}

	// Content accumulated before error must be preserved
	if s.Content() != "partial " {
		t.Errorf("content: expected 'partial ', got %q", s.Content())
	}

	// Err() returns StreamError with the error message
	err := s.Err()
	if err == nil {
		t.Fatal("expected error from Err()")
	}
	streamErr, ok := err.(*StreamError)
	if !ok {
		t.Fatalf("expected *StreamError, got %T", err)
	}
	if streamErr.Message != "provider crashed" {
		t.Errorf("expected 'provider crashed', got %q", streamErr.Message)
	}

	// Node() should fail (no done event)
	_, err = s.Node()
	if err == nil {
		t.Error("expected error from Node()")
	}
}

func TestCrossSDK_MultiLineError(t *testing.T) {
	s := streamFromFixture(fixtureMultiLineError)
	events := drainEvents(s)

	if len(events) != 2 {
		t.Fatalf("expected 2 events, got %d", len(events))
	}
	if events[0].Type != "start" {
		t.Errorf("event 0: expected start, got %s", events[0].Type)
	}

	// Multi-line data: lines joined with \n
	expectedMsg := "line one\nline two\nline three"
	if events[1].Type != "error" || events[1].Error != expectedMsg {
		t.Errorf("event 1: expected error %q, got %s %q", expectedMsg, events[1].Type, events[1].Error)
	}

	if s.Content() != "" {
		t.Errorf("content: expected empty, got %q", s.Content())
	}

	err := s.Err()
	if err == nil {
		t.Fatal("expected error from Err()")
	}
	streamErr, ok := err.(*StreamError)
	if !ok {
		t.Fatalf("expected *StreamError, got %T", err)
	}
	if streamErr.Message != expectedMsg {
		t.Errorf("expected %q, got %q", expectedMsg, streamErr.Message)
	}
}

func TestCrossSDK_ErrorOnly(t *testing.T) {
	s := streamFromFixture(fixtureErrorOnly)
	events := drainEvents(s)

	if len(events) != 1 {
		t.Fatalf("expected 1 event, got %d", len(events))
	}
	if events[0].Type != "error" || events[0].Error != "unauthorized" {
		t.Errorf("expected error 'unauthorized', got %s %q", events[0].Type, events[0].Error)
	}

	if s.Content() != "" {
		t.Errorf("content: expected empty, got %q", s.Content())
	}

	err := s.Err()
	if err == nil {
		t.Fatal("expected error from Err()")
	}
	streamErr, ok := err.(*StreamError)
	if !ok {
		t.Fatalf("expected *StreamError, got %T", err)
	}
	if streamErr.Message != "unauthorized" {
		t.Errorf("expected 'unauthorized', got %q", streamErr.Message)
	}
}

File: sdks/typescript/src/cross-sdk.test.ts
/**
 * Cross-SDK consistency tests — verify the TypeScript SDK parses SSE fixtures
 * identically to the Go and Python SDKs. The exact same SSE byte strings
 * are used in sdks/go/cross_sdk_test.go and sdks/python/tests/test_cross_sdk.py.
 */

import { describe, it, expect } from 'vitest';
import { parseSSEStream } from './sse.js';
import type { SSEEvent } from './types.js';

// Canonical SSE fixtures — byte-for-byte identical across all 3 SDKs.

const FIXTURE_NORMAL =
  'event: start\ndata: {}\n\n' +
  'event: delta\ndata: {"content":"Hello "}\n\n' +
  'event: delta\ndata: {"content":"world!"}\n\n' +
  'event: done\ndata: {"node_id":"test-node-1"}\n\n';

const FIXTURE_ERROR_MID_STREAM =
  'event: start\ndata: {}\n\n' +
  'event: delta\ndata: {"content":"partial "}\n\n' +
  'event: error\ndata: provider crashed\n\n';

const FIXTURE_MULTI_LINE_ERROR =
  'event: start\ndata: {}\n\n' +
  'event: error\ndata: line one\ndata: line two\ndata: line three\n\n';

const FIXTURE_ERROR_ONLY = 'event: error\ndata: unauthorized\n\n';

function createStream(text: string): ReadableStream<Uint8Array> {
  const encoder = new TextEncoder();
  return new ReadableStream({
    start(controller) {
      controller.enqueue(encoder.encode(text));
      controller.close();
    },
  });
}

async function collectEvents(fixture: string): Promise<SSEEvent[]> {
  const events: SSEEvent[] = [];
  for await (const event of parseSSEStream(createStream(fixture))) {
    events.push(event);
  }
  return events;
}

describe('Cross-SDK: Normal flow', () => {
  it('produces 4 events in correct order', async () => {
    const events = await collectEvents(FIXTURE_NORMAL);
    expect(events.length).toBe(4);
    expect(events[0].type).toBe('start');
    expect(events[1].type).toBe('delta');
    expect(events[2].type).toBe('delta');
    expect(events[3].type).toBe('done');
  });

  it('has correct delta content', async () => {
    const events = await collectEvents(FIXTURE_NORMAL);
    expect((events[1] as { content: string }).content).toBe('Hello ');
    expect((events[2] as { content: string }).content).toBe('world!');
  });

  it('has correct accumulated content', async () => {
    const events = await collectEvents(FIXTURE_NORMAL);
    const content = events
      .filter((e): e is { type: 'delta'; content: string } => e.type === 'delta')
      .map(e => e.content)
      .join('');
    expect(content).toBe('Hello world!');
  });

  it('has correct node_id', async () => {
    const events = await collectEvents(FIXTURE_NORMAL);
    expect((events[3] as { node_id: string }).node_id).toBe('test-node-1');
  });

  it('has no error events', async () => {
    const events = await collectEvents(FIXTURE_NORMAL);
    const errors = events.filter(e => e.type === 'error');
    expect(errors.length).toBe(0);
  });
});

describe('Cross-SDK: Error mid-stream', () => {
  it('produces 3 events in correct order', async () => {
    const events = await collectEvents(FIXTURE_ERROR_MID_STREAM);
    expect(events.length).toBe(3);
    expect(events[0].type).toBe('start');
    expect(events[1].type).toBe('delta');
    expect(events[2].type).toBe('error');
  });

  it('preserves partial content', async () => {
    const events = await collectEvents(FIXTURE_ERROR_MID_STREAM);
    expect((events[1] as { content: string }).content).toBe('partial ');
  });

  it('has correct error message', async () => {
    const events = await collectEvents(FIXTURE_ERROR_MID_STREAM);
    expect((events[2] as { error: string }).error).toBe('provider crashed');
  });

  it('has no done event', async () => {
    const events = await collectEvents(FIXTURE_ERROR_MID_STREAM);
    const doneEvents = events.filter(e => e.type === 'done');
    expect(doneEvents.length).toBe(0);
  });
});

describe('Cross-SDK: Multi-line error', () => {
  it('produces 2 events in correct order', async () => {
    const events = await collectEvents(FIXTURE_MULTI_LINE_ERROR);
    expect(events.length).toBe(2);
    expect(events[0].type).toBe('start');
    expect(events[1].type).toBe('error');
  });

  it('joins multi-line data with newlines', async () => {
    const events = await collectEvents(FIXTURE_MULTI_LINE_ERROR);
    expect((events[1] as { error: string }).error).toBe(
      'line one\nline two\nline three'
    );
  });

  it('has no delta events', async () => {
    const events = await collectEvents(FIXTURE_MULTI_LINE_ERROR);
    const deltas = events.filter(e => e.type === 'delta');
    expect(deltas.length).toBe(0);
  });
});

describe('Cross-SDK: Error only', () => {
  it('produces single error event', async () => {
    const events = await collectEvents(FIXTURE_ERROR_ONLY);
    expect(events.length).toBe(1);
    expect(events[0].type).toBe('error');
  });

  it('has correct error message', async () => {
    const events = await collectEvents(FIXTURE_ERROR_ONLY);
    expect((events[0] as { error: string }).error).toBe('unauthorized');
  });

  it('has no delta or done events', async () => {
    const events = await collectEvents(FIXTURE_ERROR_ONLY);
    const deltas = events.filter(e => e.type === 'delta');
    const dones = events.filter(e => e.type === 'done');
    expect(deltas.length).toBe(0);
    expect(dones.length).toBe(0);
  });
});
